### DISCUSSION TITLE
Transform lat/lng from collection to meilisearch _geo attribute

### DISCUSSION THREAD

**Discussion by preeco-privacy** [statamic-rad-pack/meilisearch#7] [Help] [answered]
Hi, 

in order to use meilisearch's proximity search feature i try to add the necessary _geo attribute (https://docs.meilisearch.com/learn/advanced/geosearch.html#preparing-documents-for-location-based-search)  to the documents. The statamic collection already has fields `latitude` and `longitude` but they need to be transformed to a _geo array.

My first try was something like this.. but didnt work... 

```
'transformers' => [
                'longitude' => function ($longitude) {
                    return [
                        '_geo' => [
                            'lng' => $longitude
                        ]
                    ];
                },
                'latitude' => function ($latitude) {
                    return [
                        '_geo' => [
                            'lat' => $latitude
                        ]
                    ];
                },
            ],
```

Any idea would be appreciated...

Thx,

Andreas

**Comment by tao** [comment]
My first idea would be to use a [ViewModel](https://statamic.dev/view-models) but it seems like it's not called before Search Indexing. We could ask Statamic to enable that but it might take a while. Instead we could add a config field like `geosearch => true` and then build the `_geo` field when we are preparing the default fields to be indexed. I'll work on an update this week.

**Comment by tao** [comment]
We could do something like this:

```php
        'articles' => [
            'driver' => 'meilisearch',
            'searchables' => ['collection:articles'],
            'fields' => ['title', 'content'],
            'geosearch' => true,
        ],
```

Then in the Index class add something like this:

```php
    public function insert($document)
    {
        $fields = array_merge(
            $this->searchables()->fields($document),
            $this->getDefaultFields($document),
            $this->getGeoFields($document),
        );
        $this->getIndex()->updateDocuments([$fields]);
    }
    
    private function getGeoFields(Entry|Term|LocalizedTerm|Asset|User $entry)
    {
        if ($this->config()['geosearch']) {
            return [
                'lat' => $this->get('latitude'),
                'lng' => $entry->get('longitude'),
            ];
        }
    }
```

With some kind of options to change the latitude / longitude field names and maybe format them to a specific decimal place??

It's quite a tricky problem...

**Comment by preeco-privacy** [comment, marked as answer]
Hi @tao 

thx for the inspirations! I ended up with adding a `_geo` to my Statamic collection which holds the lat and lng values separated by comma. So the transformer was easy to set up:

```
 'transformers' => [
                '_geo' => function ($_geo) {
                    $latlng = explode(',', $_geo);
                    return [
                        '_geo' => [
                            'lat' => (float) $latlng[0],
                            'lng' => (float) $latlng[1]
                        ]
                    ];
                },
```

**Reply by tao** [reply]
Great, I think long term that ViewModels would be the best solution as it allows more customisation in the future without modifying the search driver for every new change or feature that MeiliSearch makes, so I'll raise that idea with Statamic instead of changing the codebase now.
